tests/schema/tests.py

```diff
@@ -1676,6 +1676,68 @@ def test_alter_field_fk_keeps_index(self):
             {"fks": expected_fks, "uniques": 0, "indexes": expected_indexes},
         )
 
+    @skipIfDBFeature("requires_fk_constraints_to_be_recreated")
+    def test_remove_unnecessary_fk_constraint_drop(self):
+        """
+        #34417 - Tests for recreation of foreign key constraints while dropping index.
+        """
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+            editor.create_model(Book)
+
+        author = Author.objects.create(name="Alice")
+        Book.objects.create(
+            author=author,
+            title="Much Ado About Foreign Keys",
+            pub_date=datetime.datetime.now(),
+        )
+
+        old_field = Book._meta.get_field("author")
+        # dropping index on foreign key field
+        new_field = ForeignKey(Author, CASCADE, db_index=False)
+        new_field.set_attributes_from_name("author")
+        with self.assertLogs("django.db.backends.schema", "DEBUG") as cm:
+            with connection.schema_editor() as editor:
+                editor.alter_field(Book, old_field, new_field, strict=True)
+
+        # make sure foreign key constraints are not touched
+        for record in cm.records:
+            self.assertNotIn(
+                "schema_book_author_id_c80c8297_fk_schema_author_id", record.args[0]
```

Comment:
Constraint will have different name on different databases so this assertion won't work. I'd check for `DROP CONSTRAINT`.